server/src/main/java/org/elasticsearch/TransportVersion.java

```diff
@@ -109,7 +111,7 @@ private static class CurrentHolder {
 
         // finds the pluggable current version, or uses the given fallback
         private static TransportVersion findCurrent() {
-            var versionExtension = VersionExtension.load();
+            var versionExtension = ExtensionLoader.loadSingleton(ServiceLoader.load(VersionExtension.class), () -> null);
             if (versionExtension == null) {
```

Comment:
As you may have seen now, I'm a big fan of avoiding null. This is a clear enough case (null is returned just the line above) to not be really an issue, but what if we do:
```
var versionExtension = ExtensionLoader.loadSingleton(ServiceLoader.load(VersionExtension.class), 
   () -> new VersionExtensions {
      TransportVersion getCurrentTransportVersion() { 
         return TransportVersions.LATEST_DEFINED);
      }
```
or, if `VersionExtension` is a functional interface:
```
var versionExtension = ExtensionLoader.loadSingleton(ServiceLoader.load(VersionExtension.class), 
   () -> () -> TransportVersions.LATEST_DEFINED);
```
And get rid of the null check and branch?

Reply:
`VersionExtension` isn't functional, it defines the transport & index version getters. It's an interesting idea, but I'm not sure its worth going through the extra steps to define a default extension class like this, as the null is checked immediately after.

It's also probably worth keeping the different code lines for extension present & extension not present - that'll make it easy to debug/log the cases when the extension should be found but isn't. We could use `Optional` here instead maybe?

Reply:
> as the null is checked immediately after

Yes; my pet peeve with methods (and lambdas) returning null is that as a caller you never now what to expect - the contract/intention is not clear (should I handle null? what does null means? is it an error, a "value not present", a "default", etc.)
Here the _only_ usage is one line below and pretty clear, so not a big deal.
Optional usually could bring more clarity (e.g. the caller knows the value may be missing and act accordingly) but I don't think in this case is needed. I'm good with the way it is, if you think the change complicates things rather than simplify them :)